toxcore/group.c

```diff
@@ -3392,116 +3392,166 @@ uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data)
+
+        if (g->frozen == nullptr) {
+            // Memory allocation failure
+            return 0;
+        }
+    }
 
-        lendian_bytes_to_host32(&g->numfrozen, data);
-        data += sizeof(uint32_t);
+    g->title_len = *data;
 
-        if (g->numfrozen > 0) {
-            g->frozen = (Group_Peer *)calloc(g->numfrozen, sizeof(Group_Peer));
+    if (g->title_len > MAX_NAME_LENGTH) {
+        return 0;
+    }
 
-            if (g->frozen == nullptr) {
-                return STATE_LOAD_STATUS_ERROR;
-            }
+    ++data;
+
+    if (length < (uint32_t)(data - init_data) + g->title_len) {
+        return 0;
+    }
+
+    memcpy(g->title, data, g->title_len);
+    data += g->title_len;
+
+    for (uint32_t j = 0; j < g->numfrozen; ++j) {
+        if (length < (uint32_t)(data - init_data) + SAVED_PEER_SIZE_CONSTANT) {
+            return 0;
         }
 
-        g->title_len = *data;
+        Group_Peer *peer = &g->frozen[j];
+        memset(peer, 0, sizeof(Group_Peer));
 
-        if (g->title_len > MAX_NAME_LENGTH) {
-            return STATE_LOAD_STATUS_ERROR;
+        id_copy(peer->real_pk, data);
+        data += CRYPTO_PUBLIC_KEY_SIZE;
+        id_copy(peer->temp_pk, data);
+        data += CRYPTO_PUBLIC_KEY_SIZE;
+
+        lendian_bytes_to_host16(&peer->peer_number, data);
+        data += sizeof(uint16_t);
+
+        lendian_bytes_to_host64(&peer->last_active, data);
+        data += sizeof(uint64_t);
+
+        peer->nick_len = *data;
+
+        if (peer->nick_len > MAX_NAME_LENGTH) {
+            return 0;
         }
 
         ++data;
 
-        if (length < (uint32_t)(data - init_data) + g->title_len) {
-            return STATE_LOAD_STATUS_ERROR;
+        if (length < (uint32_t)(data - init_data) + peer->nick_len) {
+            return 0;
         }
 
-        memcpy(g->title, data, g->title_len);
-        data += g->title_len;
+        memcpy(peer->nick, data, peer->nick_len);
+        data += peer->nick_len;
 
-        for (uint32_t j = 0; j < g->numfrozen; ++j) {
-            if (length < (uint32_t)(data - init_data) + SAVED_PEER_SIZE_CONSTANT) {
-                return STATE_LOAD_STATUS_ERROR;
-            }
+        // NOTE: this relies on friends being loaded before conferences.
+        peer->is_friend = (getfriend_id(g_c->m, peer->real_pk) != -1);
+    }
+
+    if (g->numfrozen > g->maxfrozen) {
+        g->maxfrozen = g->numfrozen;
+    }
 
-            Group_Peer *peer = &g->frozen[j];
-            memset(peer, 0, sizeof(Group_Peer));
+    g->status = GROUPCHAT_STATUS_CONNECTED;
 
-            id_copy(peer->real_pk, data);
-            data += CRYPTO_PUBLIC_KEY_SIZE;
-            id_copy(peer->temp_pk, data);
-            data += CRYPTO_PUBLIC_KEY_SIZE;
+    id_copy(g->real_pk, nc_get_self_public_key(g_c->m->net_crypto));
 
-            lendian_bytes_to_host16(&peer->peer_number, data);
-            data += sizeof(uint16_t);
+    return (uint32_t)(data - init_data);
```

Comment:
Are all these casts safe? Do we need asserts to make sure they are?

Reply:
I just took this from the existing code, but added asserts anyway.